pkg/compose/generate.go

```diff
@@ -0,0 +1,244 @@
+*/
+
+package compose
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/compose-spec/compose-go/v2/types"
+	"github.com/docker/compose/v2/pkg/api"
+	"github.com/docker/compose/v2/pkg/utils"
+	moby "github.com/docker/docker/api/types"
+	containerType "github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/api/types/network"
+
+	"golang.org/x/exp/maps"
+)
+
+func (s *composeService) Generate(ctx context.Context, options api.GenerateOptions) (*types.Project, error) {
+	filtersListNames := filters.NewArgs()
+	filtersListIDs := filters.NewArgs()
+	for _, containerName := range options.Containers {
+		filtersListNames.Add("name", containerName)
+		filtersListIDs.Add("id", containerName)
+	}
+	containers, err := s.apiClient().ContainerList(ctx, containerType.ListOptions{
+		Filters: filtersListNames,
+		All:     true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	containersByIds, err := s.apiClient().ContainerList(ctx, containerType.ListOptions{
+		Filters: filtersListIDs,
+		All:     true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	for _, container := range containersByIds {
+		if !utils.Contains(containers, container) {
+			containers = append(containers, container)
+		}
+	}
+
+	if len(containers) == 0 {
+		return nil, fmt.Errorf("no container(s) found with the following name(s): %s", strings.Join(options.Containers, ","))
+	}
+
+	return s.createProjectFromContainers(containers)
+}
+
+func (s *composeService) createProjectFromContainers(containers []moby.Container) (*types.Project, error) {
+	project := &types.Project{}
+	services := types.Services{}
+	networks := types.Networks{}
+	volumes := types.Volumes{}
+	secrets := types.Secrets{}
+	configs := types.Configs{}
+
+	for _, c := range containers {
+		// if the container is from a previous Compose application, use the existing service name
+		serviceLabel, ok := c.Labels[api.ServiceLabel]
+		if !ok {
+			serviceLabel = getCanonicalContainerName(c)
+		}
+		service, ok := services[serviceLabel]
+		if !ok {
+			service = types.ServiceConfig{
+				Name:   serviceLabel,
+				Image:  c.Image,
+				Labels: c.Labels,
+			}
+
+		}
+		service.Scale = increment(service.Scale)
+
+		inspect, err := s.apiClient().ContainerInspect(context.Background(), c.ID)
+		if err != nil {
+			services[serviceLabel] = service
+			continue
+		}
+		s.extractComposeConfiguration(&service, inspect, volumes, secrets, networks)
+		service.Labels = cleanDockerPreviousLabels(service.Labels)
+		services[serviceLabel] = service
+	}
+
+	project.Services = services
+	project.Networks = networks
+	project.Volumes = volumes
+	project.Secrets = secrets
+	project.Configs = configs
+	return project, nil
+}
+
+func (s *composeService) extractComposeConfiguration(service *types.ServiceConfig, inspect moby.ContainerJSON, volumes types.Volumes, secrets types.Secrets, networks types.Networks) {
```

Comment:
wdyt about adding support for `cmd`, `entrypoint`, and `expose` blocks, too? they should be pretty straightforward. i'm also happy to put up a pr for those too!

Reply:
Sure, that should be part of follow up PRs

Reply:
would need to compare image CMD and ENTRYPOINT with actual container to detect those have been overridden (same for healthcheck)